src/librustc_trans/trans/base.rs

```diff
@@ -960,7 +979,9 @@ pub fn memcpy_ty<'blk, 'tcx>(bcx: Block<'blk, 'tcx>,
                              t: Ty<'tcx>) {
     let _icx = push_ctxt("memcpy_ty");
     let ccx = bcx.ccx();
-    if t.is_structural() {
+    if common::type_is_fat_ptr(bcx.tcx(), t) {
```

Comment:
@arielb1 out of curiosity, what was the red flag here that made you worry about that test case?  At first blush, I would have thought that `expr::copy_fat_ptr` will handle this correctly.